src/Tests/Microsoft.Diagnostics.Monitoring.Tool.UnitTests/CollectGCDumpActionTests.cs

```diff
@@ -0,0 +1,127 @@
+                {
+                    rootOptions.AddFileSystemEgress(ExpectedEgressProvider, uniqueEgressDirectory.FullName);
+
+                    rootOptions.CreateCollectionRule(DefaultRuleName)
+                        .AddCollectGCDumpAction(ExpectedEgressProvider)
+                        .SetStartupTrigger();
+                }, async host =>
+                {
+                    IOptionsMonitor<CollectionRuleOptions> ruleOptionsMonitor = host.Services.GetService<IOptionsMonitor<CollectionRuleOptions>>();
+                    CollectGCDumpOptions options = (CollectGCDumpOptions)ruleOptionsMonitor.Get(DefaultRuleName).Actions[0].Settings;
+
+                    ICollectionRuleActionProxy action;
+                    Assert.True(host.Services.GetService<ICollectionRuleActionOperations>().TryCreateAction(KnownCollectionRuleActions.CollectGCDump, out action));
+
+                    EndpointInfoSourceCallback callback = new(_outputHelper);
+                    await using var source = _endpointUtilities.CreateServerSource(out string transportName, callback);
+                    source.Start();
+
+                    AppRunner runner = _endpointUtilities.CreateAppRunner(transportName, TargetFrameworkMoniker.Net60); // Arbitrarily chose Net60
+
+                    Task<IEndpointInfo> newEndpointInfoTask = callback.WaitForNewEndpointInfoAsync(runner, CommonTestTimeouts.StartProcess);
+
+                    await runner.ExecuteAsync(async () =>
+                    {
+                        IEndpointInfo endpointInfo = await newEndpointInfoTask;
+
+                        using CancellationTokenSource cancellationTokenSource = new CancellationTokenSource(CommonTestTimeouts.GCDumpTimeout);
+                        CollectionRuleActionResult result = await action.ExecuteAsync(options, endpointInfo, cancellationTokenSource.Token);
+
+                        Assert.NotNull(result.OutputValues);
+                        Assert.True(result.OutputValues.TryGetValue(CollectionRuleActionConstants.EgressPathOutputValueName, out string egressPath));
+                        Assert.True(File.Exists(egressPath));
+
+                        using FileStream gcdumpStream = new(egressPath, FileMode.Open, FileAccess.Read);
+                        Assert.NotNull(gcdumpStream);
+
+                        await ValidateGCDump(gcdumpStream);
+
+                        await runner.SendCommandAsync(TestAppScenarios.AsyncWait.Commands.Continue);
+                    });
+                });
+            }
+            finally
+            {
+                try
+                {
+                    uniqueEgressDirectory?.Delete(recursive: true);
+                }
+                catch
+                {
+                }
+            }
+        }
+
+        private static async Task ValidateGCDump(Stream gcdumpStream)
+        {
+            byte[] buffer = new byte[24];
+
+            const string knownHeaderText = "!FastSerialization.1";
```

Comment:
Rather than doing the deserialization yourself, you can use `FastSerialization.Deserializer` type from `Microsoft.Diagnostics.FastSerialization` assembly. Just instantiate this type with the stream and it will parse this header out and throw an exception if that is not found.

Normally, I would recommend using the `GCHeapDump` type from `Microsoft.Diagnostics.Monitoring.EventPipe` assembly to do this, but I don't think the test assembly has the internal visibility to use it.

Reply:
I'm a bit confused about how this fits in with the other changes; assuming the use of the newly created extension method and Encoding's GetString method, what is this responsible for doing?

Reply:
Either do the other changes or do this. I was say do this one and ignore the others since it does exactly what you are doing. It probably requires the entire gcdump file to be available, which shouldn't be a problem because you've egressed it to the local file system.